### PR TITLE
More protocol friendly request pattern. Fix for #294.

### PR DESCRIPTION
**Pull request by derekcollison**
This request pattern is sematically the same but utilizes a single
wildcard subscription and the last token context for the response
subject binding to each request. Since NATS floods interest (subscriptions),
the old requestor pattern would create a new inbox subscription for each request.
It would then auto-unsubscribe and delete the subscription after the
response was received, causing quite a bit of protocol traffic.

So although this does have a performance gain as can be seen in the benchmarks,
it's implementation is specifically designed to be more friendly for NATS
clusters.

### CODE REVIEW

File: nats.go [hunk 1]
```diff
@@ -119,72 +119,72 @@ type Option func(*Options) error
...
 
-	// ClosedCB sets the closed handler called when client will
+	// ClosedCB sets the closed handler that is called when a client will
 	// no longer be connected.
 	ClosedCB ConnHandler
 
-	// DisconnectedCB sets the disconnected handler called whenever we
-	// are disconnected.
+	// DisconnectedCB sets the disconnected handler that is called
+	// whenever we are disconnected.
```
Comment (kozlovic) [inline nats.go]: Should we replaced "we are" by "the client is"?

File: nats.go [hunk 2]
```diff
@@ -217,8 +217,12 @@ type Options struct {
 	// Token sets the token to be used when connecting to a server.
 	Token string
 
-	// Dialer allows users setting a custom Dialer
+	// Dialer allows a custom Dialer when forming connections.
 	Dialer *net.Dialer
+
+	// UseOldRequestStyle force older method of Requests that utilize a new Inbox
```
Comment (kozlovic) [inline nats.go]: I see that you have changed some of the verbs for other comments, so should it be "forces" here?

File: nats.go [hunk 3]
```diff
@@ -1941,10 +1961,122 @@ func (nc *Conn) publish(subj, reply string, data []byte) error {
 	return nil
 }
 
-// Request will create an Inbox and perform a Request() call
+// respHandler is the global respnse handler. It will look up
+// the apprioriate channel based on the last token and place
```
Comment (kozlovic) [inline nats.go]: typos: "response handler" and "appropriate"

File: nats.go [hunk 4]
```diff
@@ -1941,10 +1961,122 @@ func (nc *Conn) publish(subj, reply string, data []byte) error {
...
+
+	nc.mu.Lock()
+	// Just return if closed, let Request timeout.
+	if nc.isClosed() {
+		nc.mu.Unlock()
+		return
+	}
+
+	// Grab mch
+	mch := nc.respMap[rt]
```
Comment (kozlovic) [inline nats.go]: Should we check for `nil` here? We remove the channel from the map next line, so if the responder was to send 2 responses, the second response would get a nil channel.

File: nats.go [hunk 5]
```diff
@@ -1941,10 +1961,122 @@ func (nc *Conn) publish(subj, reply string, data []byte) error {
...
+	// Delete the key regardless, one response only.
+	// FIXME(dlc) - should we track responses past 1
+	// just statistics wise?
+	delete(nc.respMap, rt)
+	nc.mu.Unlock()
+
+	// Don't block, let Request timeout instead, mch is
+	// buffered and we should delete the key before a
+	// second response is processed.
+	select {
```
Comment (kozlovic) [inline nats.go]: If we allow only 1 response per request, then why make the backing channel more than size 1, and again, since there is only 1 response (we remove the channel from the map), we should be sure to be able to put the message in the channel and not need the select, right?

File: nats.go [hunk 6]
```diff
@@ -1941,10 +1961,122 @@ func (nc *Conn) publish(subj, reply string, data []byte) error {
...
+	// _INBOX wildcard
+	ginbox := fmt.Sprintf("%s.*", NewInbox())
+	s, err := nc.Subscribe(ginbox, nc.respHandler)
+	if err != nil {
+		return err
+	}
+	// We could be racing here. So will we double check
+	// respMux here and discard the new one if set.
+	nc.mu.Lock()
+	defer nc.mu.Unlock()
```
Comment (kozlovic) [inline nats.go]: Don't use `defer` here because otherwise you will call `s.Unsubscribe()` under `nc.mu.Lock()` which will deadlock.

File: nats.go [hunk 7]
```diff
@@ -1941,10 +1961,122 @@ func (nc *Conn) publish(subj, reply string, data []byte) error {
...
+	if nc.respMux == nil {
+		nc.respSub = ginbox
+		nc.respMux = s
+		nc.respMap = make(map[string]chan *Msg)
+	} else {
+		// Discard duplicate, don't set others.
+		defer s.Unsubscribe()
```
Comment (kozlovic) [inline nats.go]: Need to explicitly release lock before calling `s.Unsubscribe()`

File: nats.go [hunk 8]
```diff
@@ -1941,10 +1961,122 @@ func (nc *Conn) publish(subj, reply string, data []byte) error {
...
+
+	// Make sure scoped subscription is setup at least once on first
+	// call to Request(). Will handle duplicates in createRespMux.
+	if doSetup {
+		if err := nc.createRespMux(); err != nil {
+			return nil, err
+		}
+	}
+	// Create literal Inbox and map to a chan msg.
+	mch := make(chan *Msg, RequestChanLen)
```
Comment (kozlovic) [inline nats.go]: Why not just size 1 here since we handle only 1 response in response handler?

**Comment by derekcollison** [comment]
I will take a look. Thought we had spell checks, etc. Good call on context, will take a look to make sure we have the appropriate tests in place, etc.

File: nats.go (hunk 7 above)
Comment (kozlovic) [inline nats.go]: By the way, your test of simultaneous requests did expose that (I was lucky that I got it the first time, but sometimes it would pass, but most times it would not on my machine).

Comment (derekcollison) [inline nats.go]: Yes my slow laptop on the plane always passed but my office machine shows it. Good catch. Will fix later tonight when the schedule frees up..

Comment (derekcollison) [inline nats.go]: And travis is even slower so they all passed too, but once on a fast machine you see it right away.

Comment (derekcollison) [inline nats.go]: At least I did write the test.. ;)

File: nats.go (hunk 8 above)
Comment (derekcollison) [inline nats.go]: paranoia around blocking channels, prefer buffered ones, even if small for safety. I also believe it changes behavior of the Go scheduler to some degree when unbuffered.

Comment (kozlovic) [inline nats.go]: I did not say unbuffered, since we would lose the message or block. I am saying why not size 1 (RequestChanLen is 8 I believe).

Comment (derekcollison) [inline nats.go]: I believe 1 or 8 is same underneath the covers memory wise.. But I could be wrong, I noticed I used 8 elsewhere and figured I picked it for a reason.

Comment (kozlovic) [inline nats.go]: It's fine, I just wanted to make sure that I understood the code correctly and that only one message should ever be put into the channel.

File: nats.go [hunk 9]
```diff
@@ -1941,10 +1961,122 @@ func (nc *Conn) publish(subj, reply string, data []byte) error {
...
+		nc.respMux = s
+		nc.respMap = make(map[string]chan *Msg)
+	} else {
+		// Discard duplicate, don't set others.
+		defer s.Unsubscribe()
+	}
+	return nil
+}
+
+// New style request that will mux on a single Subscription.
```
Comment (tylertreat) [inline nats.go]: Not true if `UseOldRequestStyle` is set.

Since this is public, seems like the docstring should describe what this does/is for rather than impl details.

Comment (tylertreat) [inline nats.go]: Also, is there a reason we need to expose the old style? There shouldn't be any compatibility issues with the switch, right? New style vs. old style is confusing from an API "UX" perspective.

Comment (derekcollison) [inline nats.go]: Good point on comment for public doc, will update.

Comment (derekcollison) [inline nats.go]: Since this is such a big change, I want to make sure clients who need to go revert back to the old way. Adding option was simple and will not affect any code unless the client is stuck for some unknown reason and wants the old way. Only situation I can think of is use of circuit breaker with 1:N where all N respond and N is large.

File: nats.go (hunk 1 above)
Comment (derekcollison) [inline nats.go]: ok

File: nats.go (hunk 2 above)
Comment (derekcollison) [inline nats.go]: yes

File: nats.go (hunk 3 above)
Comment (derekcollison) [inline nats.go]: Fixed, wonder why misspell did not catch?

File: nats.go (hunk 4 above)
Comment (derekcollison) [inline nats.go]: Sending to a nil channel will block forever, so select below will do the right thing.

File: nats.go (hunk 5 above)
Comment (derekcollison) [inline nats.go]: See above.

File: nats.go (hunk 6 above)
Comment (derekcollison) [inline nats.go]: Fixed.

File: nats.go (hunk 7 above)
Comment (derekcollison) [inline nats.go]: Fixed.

File: nats.go (hunk 9 above)
Comment (derekcollison) [inline nats.go]: Update doc, keeping ability to set option to force old style semantics.

File: context.go [hunk 10]
```diff
@@ -16,7 +16,58 @@ func (nc *Conn) RequestWithContext(ctx context.Context, subj string, data []byte
 	if ctx == nil {
 		return nil, ErrInvalidContext
 	}
+	if nc == nil {
```
Comment (kozlovic) [inline context.go]: Code coverage wise, this is not tested.

File: context.go [hunk 11]
```diff
@@ -16,7 +16,58 @@ func (nc *Conn) RequestWithContext(ctx context.Context, subj string, data []byte
...
+
+	// snapshot
+	var doSetup, useOldRequestStyle bool
+	nc.mu.Lock()
+	useOldRequestStyle = nc.Opts.UseOldRequestStyle
+	doSetup = (nc.respMux == nil)
+	nc.mu.Unlock()
+
+	// If user wants the old style.
+	if useOldRequestStyle {
```
Comment (kozlovic) [inline context.go]: Code coverage wise, this is not tested.

File: context.go [hunk 12]
```diff
@@ -16,7 +16,58 @@ func (nc *Conn) RequestWithContext(ctx context.Context, subj string, data []byte
...
+			return nil, ErrConnectionClosed
+		}
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+
+	return msg, nil
+}
 
+// oldRequestWithContext utilizes inbox and subscription per request.
```
Comment (kozlovic) [inline context.go]: This whole function shows up as not covered, which brings the code coverage of this file down to 71%.

File: nats.go [hunk 13]
```diff
@@ -1941,10 +1961,123 @@ func (nc *Conn) publish(subj, reply string, data []byte) error {
...
+		defer s.Unsubscribe()
+	}
+	nc.mu.Unlock()
+	return nil
+}
+
+// Request will send a request payload and deliver the response message,
+// or an error, including a timeout if no message was received properly.
+func (nc *Conn) Request(subj string, data []byte, timeout time.Duration) (*Msg, error) {
+	if nc == nil {
```
Comment (kozlovic) [inline nats.go]: Test for this?

File: nats.go [hunk 14]
```diff
@@ -1941,10 +1961,123 @@ func (nc *Conn) publish(subj, reply string, data []byte) error {
...
+	t := time.NewTimer(timeout)
+	defer t.Stop()
+
+	var ok bool
+	var msg *Msg
+
+	select {
+	case msg, ok = <-mch:
+		if !ok {
+			return nil, ErrConnectionClosed
```
Comment (kozlovic) [inline nats.go]: Would like to see a test where we spawn a go routine that will close the connection after a delay, and issue a request in while there is no responder running, and make sure that Request() returns ErrConnectionClosed.

**Comment by derekcollison** [comment]
@kozlovic @tylertreat @wallyqs Take a look now, should be good.

**Comment by derekcollison** [comment]
I will add some tests for RequestContext and switch to old style, and check for nil.

File: nats.go (hunk 14 above)
Comment (derekcollison) [inline nats.go]: Agree, but hard to make timing work correctly I think. I did consider it.

File: context.go (hunk 12 above)
Comment (derekcollison) [inline context.go]: Will fix..

File: nats.go (hunk 14 above)
Comment (kozlovic) [inline nats.go]: What about something like this:
```
...
wg := &sync.WaitGroup{}
wg.Add(1)
go func() {
    defer wg.Done()
    time.Sleep(250 * time.Millisecond)
    nc.Close()
}()
if _, err := nc.Request("foo", []byte("help"), 3*time.Second); err != nats.ErrInvalidConnection || err != nats.ErrConnectionClosed {
    t.Fatalf("Unexpected error: %v", err)
}
wg.Wait()
```
The test for `nats.ErrInvalidConnection` would be to avoid a flapping test where the connection would be closed in the go routine before the `Request()` call has a chance to execute.

**Comment by derekcollison** [comment]
Am I good for a merge? /cc @kozlovic @tylertreat @wallyqs

File: nats.go (hunk 14 above)
Comment (derekcollison) [inline nats.go]: That may work..

Comment (kozlovic) [inline nats.go]: I quickly tried that and looks like the Request is not kicked out...

Comment (kozlovic) [inline nats.go]: I don't think we do anything with respMux and all channels in map during close, so that makes sense...

Comment (derekcollison) [inline nats.go]: Currently this always gives us a timeout error, by design.

Comment (kozlovic) [inline nats.go]: So we would never get this ErrConnectionClosed for a closed channel, since this channel is never closed?

**Comment by tylertreat** [comment]
lgtm

File: nats.go (hunk 14 above)
Comment (derekcollison) [inline nats.go]: Fixed.

Comment (kozlovic) [inline nats.go]: Actually, if you look at Close(), we do go through the list of subscriptions and close the mch or signal in case of async. So NextMsg() (in old Request style) would get kicked out and will get ErrConnectionClosed. You are changing the behavior in that the Request will now return ErrConnectionClosed but only after waiting for the whole timeout.

File: nats.go [hunk 15]
```diff
@@ -1941,10 +1961,126 @@ func (nc *Conn) publish(subj, reply string, data []byte) error {
...
+		}
+	case <-t.C:
+		if nc.isClosed() {
```
Comment (kozlovic) [inline nats.go]: I don't think this is right (see previous comment). Also, this version of is closed is not locked.

Comment (kozlovic) [inline nats.go]: And for the test I described above (doing the close in go routine), I would also take the time before the request and after it returned and make sure that it is less than the timeout given to the request call, to make sure that the request is really kicked out ;-)

File: nats.go (hunk 14 above)
Comment (derekcollison) [inline nats.go]: Agree, fixed.

File: nats.go (hunk 15 above)
Comment (derekcollison) [inline nats.go]: Agree, fixed.

File: nats.go [hunk 16]
```diff
@@ -2746,6 +2748,20 @@ func (nc *Conn) clearPendingFlushCalls() {
...
+func (nc *Conn) clearPendingRequestCalls() {
+	if nc.respMap == nil {
+		return
+	}
+	for _, ch := range nc.respMap {
+		if ch != nil {
+			close(ch)
+		}
+	}
+	nc.pongs = nil
```
Comment (kozlovic) [inline nats.go]: did you meant nc.respMap here?

File: nats.go [hunk 17]
```diff
@@ -2068,7 +2070,7 @@ func (nc *Conn) Request(subj string, data []byte, timeout time.Duration) (*Msg,
 			return nil, ErrConnectionClosed
 		}
 	case <-t.C:
-		if nc.isClosed() {
+		if nc.IsClosed() {
```
Comment (kozlovic) [inline nats.go]: I would not put that here. It is trued that there is possibly a race that the connection is closed while the request just timed-out, but we don't do that in NextMsg, etc..

File: nats.go [hunk 18]
```diff
@@ -1968,18 +1968,20 @@ func (nc *Conn) respHandler(m *Msg) {
...
 	// Delete the key regardless, one response only.
 	// FIXME(dlc) - should we track responses past 1
 	// just statistics wise?
 	delete(nc.respMap, rt)
+
+	// Just return if closed, kick out Request by
+	// closing channel.
+	if nc.isClosed() {
+		nc.mu.Unlock()
+		close(mch)
```
Comment (kozlovic) [inline nats.go]: Is close of a nil channel valid? (remember, mch could be nil here...)

File: nats.go (hunk 17 above)
Comment (derekcollison) [inline nats.go]: ok fixed.

File: nats.go (hunk 18 above)
Comment (derekcollison) [inline nats.go]: Believe so..

Comment (derekcollison) [inline nats.go]: Actually no, but removed it anyway.

Comment (kozlovic) [inline nats.go]: No, it will panic, so this needs to be fixed.

File: nats.go [hunk 19]
```diff
@@ -2592,6 +2745,19 @@ func (nc *Conn) clearPendingFlushCalls() {
...
+// This will clear any pending Request calls.
+// Lock is assumed to be held by the caller.
+func (nc *Conn) clearPendingRequestCalls() {
+	if nc.respMap == nil {
+		return
+	}
+	for _, ch := range nc.respMap {
+		if ch != nil {
+			close(ch)
+		}
```
Comment (kozlovic) [inline nats.go]: You should empty the map. In a test, you got this:
```
github.com/nats-io/go-nats.(*Conn).respHandler(0xc420406a00, 0xc420ac21e0)
	/home/travis/gopath/src/github.com/nats-io/go-nats/nats.go:1982 +0x23a
github.com/nats-io/go-nats.(*Conn).(github.com/nats-io/go-nats.respHandler)-fm(0xc420ac21e0)
	/home/travis/gopath/src/github.com/nats-io/go-nats/nats.go:2004 +0x4c
github.com/nats-io/go-nats.(*Conn).waitForMsgs(0xc420406a00, 0xc4200f6f00)
	/home/travis/gopath/src/github.com/nats-io/go-nats/nats.go:1559 +0x1f9
created by github.com/nats-io/go-nats.(*Conn).subscribe
	/home/travis/gopath/src/github.com/nats-io/go-nats/nats.go:2223 +0x3d9
FAIL	github.com/nats-io/go-nats/test	6.891s
?   	github.com/nats-io/go-nats/util	[no test files]

Comment (kozlovic) [inline nats.go]: That's because the channel was closed, but left in the map, then respHandler tried to close again.

Comment (derekcollison) [inline nats.go]: Fixed.

File: nats.go (hunk 18 above)
Comment (kozlovic) [inline nats.go]: @derekcollison Seems that you missed above comment: closing a nil channel will panic

Comment (derekcollison) [inline nats.go]: Reverted back to just returning.

**Comment by kozlovic** [comment]
LGTM!

**Comment by derekcollison** [comment]
Sorry that was so much a PITA.. old and rusty, uug.
